Warn on instantiating records w/ ambiguous columns

khasinski opened the pull request:
### Summary

It is possible to query for models with a result that has duplicate
columns. For example:

```ruby
Post.joins(:author).select("*").first
```

will have at least two `id` columns (and possibly more).

Loading record matches the column names and will overwrite
the resulting column several times on duplicates which might
lead to unexpected query results.

Fixes #41151


### Other Information

I'd love to have a better warning message and some feedback if this is the right place to issue this warning.

I've decided to check for duplicates in-line instead of adding another method as it seems `ActiveRecord::Querying` is intentionally slim. I also decided against any clever way of finding duplicate names faster than O(n^2) since the number of columns isn't usually large.

khasinski commented:
Bump, since the ticket is still open

p8 commented on activerecord/test/cases/associations/join_model_test.rb:
```diff
@@ -772,6 +772,12 @@ def test_proper_error_message_for_eager_load_and_includes_association_errors
+  def test_warn_if_duplicate_column_names_detected
+    assert_called(ActiveRecord::Base.logger, :warn) do
+      Post.joins(:author).select("*").first
+    end
```
Could you also assert the warning message contains the expected message?

khasinski commented:
I'm not sure why the build is failing, but I've rebased it on top of `main` since it was quite a lot behind the current version.

rafaelfranca commented on activerecord/lib/active_record/querying.rb:
```diff
@@ -69,6 +69,11 @@ def _load_from_sql(result_set, &block) # :nodoc:
 
+      duplicate_column_names = result_set.columns.tally.select { |_, v| v > 1 }.keys
+      if duplicate_column_names.any?
+        logger.warn "Duplicate column names: #{duplicate_column_names.join(', ')}"
```
`logger` can be `nil`, so we need to check it here.

rafaelfranca commented on activerecord/lib/active_record/querying.rb:
```diff
@@ -69,6 +69,11 @@ def _load_from_sql(result_set, &block) # :nodoc:
 
+      duplicate_column_names = result_set.columns.tally.select { |_, v| v > 1 }.keys
+      if duplicate_column_names.any?
+        logger.warn "Duplicate column names: #{duplicate_column_names.join(', ')}"
```
This message isn't actionable. And the SQL can be executed very far away form the code that generated it, so we need to include which SQL has the duplicated names. In fact, I'm not sure a warning is the right thing to do here. Is there any case were this behavior is intended and not a mistake?

khasinski replied:
This behavior _might_ be intended (in terms of user who can explicitly name two columns the same), but results in a model that might have odd properties. I'd say the most common problem is when you have multiple identity columns, since you can inadvertently fetch a record, update it and save it with a wrong `id`, overwriting some data.

I've stumbled upon this behaviour in client's app code that's why I'm interested in at least logging this case. Perhaps we can just add a warning for multiple identity columns? 

Also - any suggestions for better error message?

khasinski commented:
This got stale (and had some conflicts 🤔), so I've rebased it on top of `main`

zzak commented:
There are a bunch of failures in CI, would you mind taking a look?

khasinski commented:
@zzak all done

khasinski committed:
Warn on instantiating records w/ ambiguous columns

It is possible to query for models with a result that has duplicate
columns. For example

```ruby
Post.joins(:author).select("*").first
```

Will have at least two `id` columns (and possibly more).

Loading record matches the column names and will overwrite
the resulting column several times on duplicates which might
lead to unexpected query results.

Fixes #41151

khasinski committed:
Fix test to work when logger is nil

On CI (Buildkite), ActiveRecord::Base.logger is set to nil.
The test needs to explicitly set a logger to verify the warning
is emitted when duplicate column names are detected.
